src/main/java/nl/knaw/dans/dvcli/command/NotificationTruncate.java

```diff
@@ -0,0 +1,161 @@
+        int user; // a number, preventing accidental SQL injection
+        // This id is visible for 'superusers' in the Dataverse Dashboard
+        
+        @CommandLine.Option(names = { "--all-users" }, required = true, 
+                description = "Truncate notifications for all users.")
+        boolean allUsers;
+    }
+    
+    @CommandLine.Parameters(index = "0", paramLabel = "number-of-records-to-keep", 
+            description = "The number of notification records to keep.")
+    private int numberOfRecordsToKeep;
+    
+    private record NotificationTruncateParams(Database db, int userId, int numberOfRecordsToKeep) {
+    }
+
+    protected <NotificationTruncateParams> BatchProcessor.BatchProcessorBuilder<NotificationTruncate.NotificationTruncateParams, String> paramsBatchProcessorBuilder() throws IOException {
+        return BatchProcessor.<NotificationTruncate.NotificationTruncateParams, String> builder();
+
+    }
+    
+    private static class NotificationTruncateAction implements ThrowingFunction<NotificationTruncate.NotificationTruncateParams, String, Exception> {
+
+        @Override
+        public String apply(NotificationTruncateParams notificationTruncateParams) throws Exception {
+            // Dry-run; show what will be deleted
+            //List<List<String>> results = notificationTruncateParams.db.query(String.format("SELECT * FROM usernotification WHERE user_id = '%d' AND id NOT IN (SELECT id FROM usernotification WHERE user_id = '%d' ORDER BY senddate DESC LIMIT %d);", 
+            //        notificationTruncateParams.userId, notificationTruncateParams.userId, 
+            //        notificationTruncateParams.numberOfRecordsToKeep), true
+            //);
+            //return "Notifications for user " + notificationTruncateParams.userId + " that will be deleted: \n" 
+            //        + getResultsAsString(results);
+            
+            // Actually delete the notifications
+            try {
+                log.info("Deleting notifications for user with id " + notificationTruncateParams.userId);
+                int rowCount = notificationTruncateParams.db.update(String.format("DELETE FROM usernotification WHERE user_id = '%d' AND id NOT IN (SELECT id FROM usernotification WHERE user_id = '%d' ORDER BY senddate DESC LIMIT %d);",
+                        notificationTruncateParams.userId, notificationTruncateParams.userId,
+                        notificationTruncateParams.numberOfRecordsToKeep));
+                return "Deleted " + rowCount + " record(s) for user with id " + notificationTruncateParams.userId;
+            } catch (SQLException e) {
+                throw new Exception("Error deleting notifications for user with id " + notificationTruncateParams.userId, e);
+            }
+        }
+    }
+    
+    @Override
+    public void doCall() throws Exception {
+        // validate input
+        if (numberOfRecordsToKeep < 0) {
```

Comment:
Annotate the field with positive

Reply:
How, cannot find this in picocli docs?

Reply:
It's not doing anything, I have to test and throw an exception myself. What am I doing wrong?

Reply:
Ok, it's only working within the dropwizzard framework so we will just keep validating ourselfes